### PR TITLE
Add bulk user creation page

### PR DESCRIPTION
**Pull request by oliverviljamaa**
## Context

We need to add a page for admin users to be able to create users with a certain role.

## Changes

A _bulk order creation page_ is added under the `/admin/create-users` URL, accessible only to users with the `BULK_CREATE_USERS` permission.

## Considerations

Dependent on https://github.com/cov-clear/backend/pull/65.

### CODE REVIEW

File: src/admin/BulkUserCreationPage.test.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+jest.mock('axios');
+jest.mock('../authentication/context');
+
+describe('Bulk user creation page', () => {
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('creates users with emails from comma-separated input and selected role and shows a success or error message', async () => {
```
Comment (oliverviljamaa) [inline src/admin/BulkUserCreationPage.test.tsx:17]: It's a big test, but with React Testing Library's integration-rather-than-unit style, I've found longer tests to be acceptable, as we reduce repetition, over-abstraction, and still manage to find out what went wrong easily thanks to its great logging.

File: src/admin/BulkUserCreationPage.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import { Box, Label, Heading, Container, Button, Text, Textarea, Alert, Select } from 'theme-ui';
+import { useFormik } from 'formik';
+import * as yup from 'yup';
+
+import { Role, CreateUserCommand } from '../api';
+import useBulkUserCreation from './hooks/useBulkUserCreation';
+
+const AnyBox = Box as any;
+
+const ROLES: Role[] = ['TRUSTED_PATIENT', 'CLINICIAN'];
```
Comment (oliverviljamaa) [inline src/admin/BulkUserCreationPage.tsx:11]: What should the role names be?

File: src/admin/BulkUserCreationPage.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }),
+    onSubmit: ({ emailsString, role }) => {
+      const emails = getEmails(emailsString);
+      const command: CreateUserCommand[] = emails.map((email) => ({ email, roles: [role] }));
+      create(command);
+    },
+  });
+
+  return (
+    <Container variant="page">
```
Comment (oliverviljamaa) [inline src/admin/BulkUserCreationPage.tsx:33]: As this is a back-office page, it doesn't need to follow the mobile-specific layout, but I also saw no harm in it.

File: src/admin/BulkUserCreationPage.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      <Heading as="h1" mb={4}>
+        Create users
+      </Heading>
+
+      <Text mb={4}>Users with the emails will be created with the selected role.</Text>
```
Comment (oliverviljamaa) [inline src/admin/BulkUserCreationPage.tsx:38]: Any copy improvement ideas?

File: src/api.ts [hunk 5]
```diff
@@ -97,6 +97,21 @@ export async function updateUser(user: User, options: AuthenticatedHttpOptions):
...
+export interface CreateUserCommand {
+  email: string;
+  roles: Role[];
+}
+
+export async function createUsers(
+  command: CreateUserCommand[],
+  options: AuthenticatedHttpOptions
+): Promise<User[]> {
+  const response = await authenticated(options.token).post('/api/v1/users', command);
```
Comment (oliverviljamaa) [inline src/api.ts:111]: Do we need the cancel token functionality here? I see it being supported for some queries, but not being used.

File: src/theme.ts [hunk 6]
```diff
@@ -91,6 +91,12 @@ const theme = {
       paddingLeft: 4,
       paddingRight: 4,
     },
+    success: {
+      background: '#52c41a', // Primary green from https://ant.design/docs/spec/colors
```
Comment (oliverviljamaa) [inline src/theme.ts:95]: Went with changes as minimal as possible here.

File: src/admin/BulkUserCreationPage.tsx (hunk 2 above)
Comment (kstam) [reply src/admin/BulkUserCreationPage.tsx:11]:
```
GET /api/v1/roles
```
 😄

Comment (kstam) [reply src/admin/BulkUserCreationPage.tsx:11]: What I mean is that they should come from the API and not get hard-coded here. We haven't defined them yet on the backend.

Comment (oliverviljamaa) [reply src/admin/BulkUserCreationPage.tsx]: ✅ The roles are now fetched from the API.

File: src/admin/hooks/useRoles.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  error: NullableError;
+  roles: Role[];
+} {
+  const { token } = useAuthentication();
+
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState(null as NullableError);
+  const [roles, setRoles] = useState([] as Role[]);
+
+  useEffect(() => {
```
Comment (oliverviljamaa) [inline src/admin/hooks/useRoles.ts:19]: Once https://github.com/cov-clear/web/pull/25 is merged, I'll rebase and make use of the common `useAuthenticatedHttpResource` to reduce duplication.

File: src/admin/BulkUserCreationPage.tsx (hunk 4 above)
Comment (kstam) [reply src/admin/BulkUserCreationPage.tsx:51]: Not sure if it's better. 
```
Creates users with the specified emails and assigns the selected role to them. 

Note: If a user already exists for a given email, the selected role will be assigned to them.
```

File: src/admin/BulkUserCreationPage.test.tsx [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    await waitFor(() => {
+      expect(getByText('Failed to load roles.')).toBeInTheDocument();
+    });
+  });
+
+  function mockAuthentication() {
+    (useAuthentication as jest.Mock).mockReturnValue({ token: 'some-token' });
+  }
+
+  function anUser(email: User['email']): User {
```
Comment (Tankenstein) [inline src/admin/BulkUserCreationPage.test.tsx:99]:
```suggestion
  function aUser(email: User['email']): User {
```

File: src/admin/BulkUserCreationPage.test.tsx [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    const roleSelect = getByLabelText(/role/i);
+
+    await waitFor(() => {
+      expect(getByText('USER')).toBeInTheDocument();
+      expect(getByText('DOCTOR')).toBeInTheDocument();
+    });
+
+    mockPost.mockResolvedValueOnce({
+      data: [
+        anUser('jane.doe@example.com'),
```
Comment (Tankenstein) [inline src/admin/BulkUserCreationPage.test.tsx:45]:
```suggestion
        aUser('jane.doe@example.com'),
```

File: src/admin/BulkUserCreationPage.test.tsx [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        anUser('john.doe@gmail.com'),
```
Comment (Tankenstein) [inline src/admin/BulkUserCreationPage.test.tsx:46]:
```suggestion
        aUser('john.doe@gmail.com'),
```

File: src/admin/BulkUserCreationPage.test.tsx [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        anUser('johnny.doe@yahoo.com'),
```
Comment (Tankenstein) [inline src/admin/BulkUserCreationPage.test.tsx:47]:
```suggestion
        aUser('johnny.doe@yahoo.com'),
```

File: src/admin/hooks/useRoles.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
+import { useState, useEffect } from 'react';
```
Comment (Tankenstein) [inline src/admin/hooks/useRoles.ts:1]: I don't find it very useful to separate hooks into `hooks` directory, but it's probably just personal preference.

File: src/admin/hooks/useRoles.ts (hunk 7 above)
Comment (Tankenstein) [reply src/admin/hooks/useRoles.ts:19]: Was going to add a comment about cancellation but that'll take care of it i guess

File: src/admin/BulkUserCreationPage.tsx [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    validationSchema: yup.object().shape({
+      role: yup.string().required(),
+      emailsString: yup.string().required(),
+    }),
+    onSubmit: ({ role, emailsString }) => {
+      const emails = getEmails(emailsString);
+      const command: CreateUserCommand[] = emails.map((email) => ({ email, roles: [role] }));
+
+      if (
+        window.confirm(
```
Comment (Tankenstein) [inline src/admin/BulkUserCreationPage.tsx:36]: nice

File: src/api.ts (hunk 5 above)
Comment (Tankenstein) [reply src/api.ts:109]: I'd support it. If you use the common hook later it'll be used. Also, you still probably want cancellation in your hook above.

File: src/admin/BulkUserCreationPage.tsx (hunk 13 above)
Comment (oliverviljamaa) [reply src/admin/BulkUserCreationPage.tsx:36]: Seemed dangerous, thought I'd throw it in.

File: src/admin/hooks/useRoles.ts (hunk 12 above)
Comment (oliverviljamaa) [reply src/admin/hooks/useRoles.ts]: ✅ Agreed, removed the wrapper directory.

File: src/admin/BulkUserCreationPage.tsx (hunk 4 above)
Comment (oliverviljamaa) [reply src/admin/BulkUserCreationPage.tsx]: Thanks, went with a slightly modified version:
```
Creates users with the specified emails and assigns the selected role to them.
If a user with a given email already exists, the selected role will be assigned to them.
```

File: src/api.ts (hunk 5 above)
Comment (oliverviljamaa) [reply src/api.ts:109]: Will add it when we'll generalise the common hook to support creation in addition to retrieval as well (at the moment, it fetches on initialisation, so needs a few changes).

**Pull request merged**